feat: setup RTL support

TylerAPfledderer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Sets up RTL support as described in the [Chakra Docs on RTL-aware style props](https://chakra-ui.com/docs/styled-system/rtl-support#using-rtl-aware-style-props)
<!--- Describe your changes in detail -->

- Declare language direction globally
  - Move `extendBaseTheme` to `_app.tsx` to allow for the `direction` them prop to be defined with a locale check from the router.
  - Convert `_document.tsx` to a class in order to access the locale and set up the `dir` attribute on `Html`
- Remove `dir` prop declarations from the page layout components
- Make some noticeable adjustments with RTL-aware style props

## Related Issue
Closes #78 
<!--- This project accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

TylerAPfledderer committed:
feat: setup RTL support

TylerAPfledderer commented on src/pages/_document.tsx:
```diff
@@ -1,13 +1,32 @@
+} from "next/document"
+
+class Document extends NextDocument {
+  static async getInitialProps(ctx: DocumentContext) {
+    return await NextDocument.getInitialProps(ctx)
+  }
```
Per [response to the GH Issue](https://github.com/ethereum/ethereum-org-fork/issues/78#issuecomment-1793688702) and my follow-up, maybe this can simply be removed? I have not found an issue locally when this method override is not applied. But need to be allowed to check the preview deploy to see if anything goes wrong after build.

```suggestion
```

wackerow commented on src/components/TranslationBanner.tsx:
```diff
@@ -40,7 +40,7 @@ const TranslationBanner: React.FC<IProps> = ({
       display={isOpen ? "block" : "none"}
       bottom={{ base: 0, md: 8 }}
-      right={{ base: 0, md: 8 }}
+      insetInlineEnd={{ base: 0, md: 8 }}
```
Curious what the scope of the PR is intended to be... I see we're starting to update some of the properties to be direction-aware, but a quick search shows many more to tackle

Just searching `marginRight`:
<img width="391" alt="image" src="https://github.com/ethereum/ethereum-org-fork/assets/54227730/796e598a-9b32-4449-8fc5-310211c815e3">

Am I correct that these still (and others similar like `mr` `ml` `pr` `pl` `borderLeft/Right` etc). need to be handled? Or is there a reason I'm missing that we don't need to?

cc: @pettinarip

TylerAPfledderer replied:
@wackerow yep, that was to be expected. I only made updates to those that were very noticeable.

I don't expect everything to be directionally aware in design, but I can go ahead and review in this PR the components with props that should be flipped like what you showed.

@pettinarip do advise still your thoughts and if this should be handled here. 😁

pettinarip replied:
@wackerow good observation. Probably I wasn't clear enough but the idea is to do the initial setup for rtl support in the nextjs repo by using chakra.

I agree that doing the entire adaptation now would be scope creep. That is why I think we should do all that heavy lifting after we finish the migration, as a separate effort.

Do you see some other issues with doing this in two parts?

wackerow replied:
>  agree that doing the entire adaptation now would be scope creep.

Agree as well... just thinking. Adding `dir="rtl"` to the `html` tag will flip the entire layout... two items next to each other will be flipped, and any `mr/marginRight` etc. being used is unlikely to look correct anymore.

A small example, I went to https://ethereum.org/ar/nft/ and manually added `dir="rtl"` to the html tag in the DOM:

<img width="1326" alt="image" src="https://github.com/ethereum/ethereum-org-fork/assets/54227730/67d3fe5a-e4aa-415c-b2a4-7341396410a3">

The icon in this banner has a "margin right" on it... which only works appropriately when it is to the left of the text... in the image above you see the gap between the icon and the text is gone. 

---

I'm not sure exactly how many of these will surface with this switch, but overall I don't think it will be all that serious in the short term, and think it's probably best to prevent the scope creep, get this PR in without being blocked by updating everything, and splitting that as a near-future task.

TylerAPfledderer committed:
Merge branch 'dev' into feat/rtl-support

TylerAPfledderer committed:
fix: update changes from 'dev' merge

TylerAPfledderer committed:
Merge branch 'dev' into feat/rtl-support

wackerow committed:
update all shorthand margin/padding directionals

ml=>ms, mr=>me, pl=>ps, pr=>pe, mx/px where applicable

wackerow committed:
update all margin/padding directionals

marginRight=>me, marginLeft=>ms, paddingRight=>pe, paddingLeft=>ps

wackerow committed:
update absolute left/right positioning

left=>insetInlineStart, right=>insetInlineEnd

wackerow committed:
update left/right border directionals

borderRight=>borderInlineEnd, borderLeft=>borderInlineStart, borderTopLeftRadius=>borderTopStartRadius, borderTopRightRadius=>borderTopEndRadius, etc

wackerow committed:
update margin/padding directional

wackerow committed:
update alignments to use start/end

wackerow committed:
create and implement useRtlFlip hook

Applied to elements that should be visually flipped for RTL languages

wackerow committed:
patch missing data-group for groupHover

wackerow committed:
Merge branch 'dev' into pr/112

wackerow committed:
patch EnergyConsumptionChart

recharts Margin type requires use of right/left, but inconsequential as all are equal; add isRtl check for data order within bar chart

wackerow committed:
apply flipForRtd to arrow characters

ie: ↗ and ↘

wackerow committed:
remove redundant `dir` declarations

in lieu of top-level html dir attribute

wackerow reviewed (approved):
Spiked on updating as many style props I could find that are right/left oriented and needed to be updated for RTL responsiveness.

Think I got all the main ones, but the trickier RTL stuff involves things like directional arrows... 
Added a `useRtlFlip` hook (open to potential other names here, but this seemed to work) that will provide a `transform` string value to horizontally flip/mirror an element if it's an RTL locale. This fixed things like the arrow on the DocLink component, the external links, or the right-angle decorators in the Table of Contents.

This should cover the majority of cases, but there are definitely more we'll discover... ie, one I just found in the dev docs sidebar:
<img width="401" alt="image" src="https://github.com/ethereum/ethereum-org-fork/assets/54227730/484c799f-5cb1-438a-8d58-54fdd3503265">

...Chevron's are pointing the wrong way (so we could apply the hook above to this case). 

Going to approve this now; I think it's worth just bringing this in and we can continue to clean up lose ends on a lower priority basis after the critical pieces of the migration are completed. 

Solid work @TylerAPfledderer! This should be a huge improvement for RTL locales 🎉

wackerow committed:
add TODO

pettinarip reviewed (approved):
This is great! thanks @TylerAPfledderer & @wackerow

One thing that I think we should also consider in the future is related to non-translated pages, like https://deploy-preview-112--ethereum-org-fork.netlify.app/ar/web3

In that case, I think we shouldn't change the direction, at least, of the content. The rest I think it is ok.
